File: Merge_sort.py
import random
import time
import sys

# --- MERGE SORT IMPLEMENTATION ---
def merge_sort(arr):
    """
    Recursively splits and merges the array.
    """
    if len(arr) > 1:
        mid = len(arr) // 2
        left_half = arr[:mid]
        right_half = arr[mid:]

        merge_sort(left_half)
        merge_sort(right_half)

        i = j = k = 0
        # Merge the two halves
        while i < len(left_half) and j < len(right_half):
            if left_half[i] < right_half[j]:
                arr[k] = left_half[i]
                i += 1
            else:
                arr[k] = right_half[j]
                j += 1
            k += 1

        # Copy the remaining elements of left_half, if any
        while i < len(left_half):
            arr[k] = left_half[i]
            i += 1
            k += 1

        # Copy the remaining elements of right_half, if any
        while j < len(right_half):
            arr[k] = right_half[j]
            j += 1
            k += 1

def main():
    """
    Usage: python merge_sort.py <n>
      - n is the size of the array to sort. Default = 1e6 (1 million).
    """
    if len(sys.argv) > 1:
        n = int(sys.argv[1])
    else:
        n = 10**6  # Default size if no argument is given

    print(f"Generating array of size {n} ...")
    arr = [random.randint(0, 10**9) for _ in range(n)]

    print("Starting merge sort...")
    start_time = time.time()
    merge_sort(arr)
    end_time = time.time()

    print(f"Time taken to sort {n} elements: {end_time - start_time:.4f} seconds")

    # Optional: check if array is sorted
    # (This can be expensive to do for large n, so you may remove this check.)
    assert all(arr[i] <= arr[i+1] for i in range(n-1)), "Array is not sorted!"

if __name__ == "__main__":
    main()

File: merge_sort.cpp
#include <iostream>
#include <vector>
#include <cstdlib>
#include <ctime>
#include <algorithm>  // For std::swap, if needed
#include <chrono>

using namespace std;

// --- MERGE SORT IMPLEMENTATION ---
void merge(vector<int> &arr, int left, int mid, int right) {
    int n1 = mid - left + 1;
    int n2 = right - mid;

    vector<int> L(n1), R(n2);

    for(int i = 0; i < n1; i++) {
        L[i] = arr[left + i];
    }
    for(int j = 0; j < n2; j++) {
        R[j] = arr[mid + 1 + j];
    }

    int i = 0, j = 0, k = left;
    while(i < n1 && j < n2) {
        if(L[i] <= R[j]) {
            arr[k++] = L[i++];
        } else {
            arr[k++] = R[j++];
        }
    }
    while(i < n1) {
        arr[k++] = L[i++];
    }
    while(j < n2) {
        arr[k++] = R[j++];
    }
}

void mergeSort(vector<int> &arr, int left, int right) {
    if(left < right) {
        int mid = left + (right - left) / 2;
        mergeSort(arr, left, mid);
        mergeSort(arr, mid + 1, right);
        merge(arr, left, mid, right);
    }
}

int main(int argc, char** argv) {
    // Default array size is 1e6 if no argument is provided
    long long n = 1000000;
    if(argc > 1) {
        n = atoll(argv[1]);
    }

    cout << "Generating array of size " << n << " ..." << endl;
    srand((unsigned) time(NULL));
    vector<int> arr(n);
    for(long long i = 0; i < n; i++) {
        arr[i] = rand(); // random int
    }

    cout << "Starting merge sort..." << endl;
    auto start = chrono::high_resolution_clock::now();
    mergeSort(arr, 0, n - 1);
    auto end = chrono::high_resolution_clock::now();
    chrono::duration<double> diff = end - start;

    cout << "Time taken to sort " << n << " elements: " 
         << diff.count() << " seconds" << endl;

    // Optional: verify that arr is sorted
    // for(int i = 0; i < n - 1; i++){
    //     if(arr[i] > arr[i+1]){
    //         cerr << "Array not sorted!" << endl;
    //         break;
    //     }
    // }

    return 0;
}
